### PR TITLE
Fix syntax disambiguation for declaration expressions and the is operator

### PR DESCRIPTION
**Pull request by gafter**
Modifies the disambiguation in a tuple literal so that a declaration expression must be followed by a `,` or `)`, and modifies the disambiguation in an `is` expression so existing code is not affected.

**Customer scenario**

We "broke" code that worked correctly previously. For example, in the statement

``` c#
var j = a < i >> 2;
```

We were treating `a<i>` as a generic (incorrect), rather than as a comparison with a shift (correct). This bug breaks existing (pre c#7) code.

Also, in the statement
``` c#
return (i, isValid ? Errors.IsValid : Errors.HasErrors);
```
we were incorrectly interpreting `isValid?` as a nullable type, and `isValid? Errors` as a declaration expression. The rest of the stuff before the close paren then appeared to be garbage by the parser.
This PR adjusts the disambiguation so that we only parse a declaration expression in this context when it is followed by a `,` or `)`.

**Bugs this fixes:** 

Fixes https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=377111
Fixes https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=377556

**Workarounds, if any**

It is possible to "fix" customer code broken by these bugs by parenthesizing the `?:` conditional expression (first issue) or parenthesizing the shift operator on the right-hand-side of the comparison (second issue). Unfortunately, some real customer code bases are affected by this, and it is not clear from the compiler error messages what went wrong or how the customer should fix it. Also, this breaks existing customer code that compiled without problem before C# 7, so this could be an adoption blocker.

**Risk**

Low; this fix narrows the disambiguation code so that it only accepts declaration expressions and parses type arguments in more appropriate contexts.

**Performance impact**

None, as the amount of code executed for typical workloads is the same (it is merely a change to the logic executed).

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

This was a fallout from a recent complete overhaul for the parsing of patterns and declaration expressions, which are new features in C# 7. We did not have tests for these particular intersections of syntactic features. It is possible that this would have been caught in testing had the compiler team followed through on its plan to assign a team member to adversarial testing of new features. 

**How was the bug found?**

Customer reported (these bugs broke code that worked previously)

@jcouv @VSadov Please review.

### CODE REVIEW

**Comment by gafter** [comment]
@jcouv I don't know a "lightweight" way to do it. If I were assigned the adversarial test role for any of these features, I would probably aim to shake these out by writing a test generator (that verifies the parse tree against the shape of the generated syntax).

**Comment by gafter** [comment]
I've added a fix for [https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=377556](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=377556) to this PR. I will update the description above shortly.

@jcouv Can you please review the changes since your last review, for the new bug fix?
@VSadov Can you please review this change for possible integration in RTM?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -6451,7 +6460,8 @@ private enum ParseTypeMode
         {
             Normal,
             Parameter,
-            AfterIsOrCase,
+            AfterIs,
+            AfterCase,
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Do we parse differently after "is" and after "case" ?

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I see, we do now. Used to pass an extra bool.

---
In reply to: [98789489](https://github.com/dotnet/roslyn/pull/16834#discussion_r98789489) [](ancestors = 98789489)

**Comment by VSadov** [comment]
:+1:

**Comment by gafter** [comment]
@jcouv Can you please review the changes since your last review, for the new bug fix? I am proposing to get this into RTM.

**Comment by jcouv** [comment]
I'm on it. Sorry for the delay.

**Comment by jcouv** [comment]
@gafter I've been thinking about how to cover more scenarios with smaller tests. A lightweight verification would just check a few nodes. For instance, I expect one generic type name, one shift operator and no missing nodes. Or in the tuple case, I expect no missing nodes, and one conditional operator.

**Comment by jcouv** [comment]
Also reported in https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=393562

**Pull request merged**